### DISCUSSION TITLE
How to add Single Entry FinderComment

### DISCUSSION THREAD

**Discussion by PDHell** [RhetTbull/osxphotos#1529] [Q&A]
Hi,

First of all many many thanks for writing such a fantastic tool to export photos. 
I can't wait for your GUI tool. 

I have one small question, please see if you can answer. 
Here is my osxphotos command:

```
osxphotos \
export     ~/Pictures/OSXPhotos_Backup \
--library  ~/Pictures/Photos\ Library.photoslibrary \
--filename "{created.date}.{original_name}{favorite?.Fav,}" \
--report   ~/Pictures/OSXPhotos_Backup/exportlog_{today.date}_{today.hour}H.{today.min}M.{today.sec}S.csv \
--save-config ~/Pictures/OSXPhotos_Backup/osxphotos.toml \
--update \
--cleanup \
--export-by-date \
--touch-file \
--ramdb  \
--verbose \
--not-hidden \
--keyword-template "{label,KarNoLabel}" \
--keyword-template "{folder_album}" \
--person-keyword \
--xattr-template findercomment "Persons = {person} ; Place = {place.name}; Labels = {label}" \
--keep "/*.csv"  \
--keep "/*.toml"  \
--sidecar XMP
```

I am trying to achive:
1.  Copy "PhotoLibrary" photos into finder without taking additional space (hence not using `--exiftool`)
2. Also, Dump all Metadata info to XML.
3. And, dump "Pesons" and "Labels/Keywords" into FinderComments -- which can help in spotlight searches. 

Everything working perfectly except "FinderComments". Below are my FinderComments to one of my photo:
```
Persons = Person1 ; Place = San Jose, California, United States; Labels = Child, Persons = Person2 ; 
Place = San Jose, California, United States; Labels = Child, Persons = Person1 ; Place = San Jose, California, United States;
Labels = People, Persons = Person2 ; Place = San Jose, California, United States; Labels = People, 
Persons = Person1 ; Place = San Jose, California, United States; Labels = Selfie, Persons = Person2 ; 
Place = San Jose, California, United States; Labels = Selfie
```

As you can see I have 2 persons, 1 place, 3 label in this photos. But because `Person` and `label` result in multiple values, it is getting multiplied and all fileds are displaying 3 times. Is it possible to have finder comment like below?
``Persons = Person1, Person2; Place = San Jose, California, United States; Labels = Child"``

Thanks,

**Comment by RhetTbull** [comment]
This is because the template system works in a a combinatorial fashion -- for templates that have multiple values, like `person` and `label`, the template generates a string with every combination of each of the values.  This is a unique feature of the OSXPhotos template system and one of the reasons I built my own instead of using an existing systems. This was needed to handle exporting to multiple directories or filenames depending on templates.  But there is a way  (two actually) to get the behavior you want:

You can tell the template system to join all the values of a single template together using a delimiter by prepending the delimiter and `+` sign before the template field (in this case we'll use comma + space as delimiter:

`Persons = {, +person} ; Place = {place.name} ; Labels = {, +label}`

This produces output like this:

`Persons = Katie, Suzy ; Place = Omaha, Nebraska, United States ; Labels = Agriculture, Blue Sky, Boot, Child, Clothing, Farm, Food, Footwear, Outdoor, People, Pumpkin, Shoes, Sky, Vegetable`

This can also be accomplished using the `join()` filter:

`Persons = {person|join( ,)} ; Place = {place.name}; Labels = {label|join(, )}`

I prefer the first because it's more succinct but both accomplish the same thing.

There is one issue with this template in that if any of these fields are empty (e.g. no person in the photo or no place name), you'll get something that looks like this:

`Persons = _ ; Place = St James's Park, London, England, United Kingdom; Labels = Cloudy, Land, Outdoor, Plant, River, Sky, Water, Water Body, Waterways`

Where it still shows the heading `Persons = ` but the value `_` which is the default value for null in the template system (this is due to the fact that the template must produce a valid filename as that is one of the primary uses).  You can leave this if it doesn't bother you or provide your own default. For example:

`Persons = {, +person,None} ; Place = {place.name,None} ; Labels = {, +label,None}`

Produces

`Persons = None ; Place = St James's Park, London, England, United Kingdom; Labels = Cloudy, Land, Outdoor, Plant, River, Sky, Water, Water Body, Waterways`

If preferred, you can also use the conditional test feature (`{field?`) of the template system to eliminate the header if the value is null:

`{person?Persons = {, +person} ; ,}{place.name?Place = {place.name} ; ,}{label?Labels = {, +label},}`

This yields:

`Place = St James's Park, London, England, United Kingdom ; Labels = Cloudy, Land, Outdoor, Plant, River, Sky, Water, Water Body, Waterways`

Note that the `Persons =` was dropped because there were no people in the photo. We use a null default value `,}` so that if the condition is false (no person) then no value is produced by the template.

Hope this helps!

**Reply by RhetTbull** [reply]
By the way, the `osxphotos template` command is useful for constructing and testing templates in an interactive way. Select a photo in Photos and launch the the command then you can try out different templates and they'll be rendered for the selected photo. If you change the selection, the template is applied to the new photo.
